### DISCUSSION TITLE
Scala Native 0.5 plans

### DISCUSSION THREAD

**Discussion by rossabaker** [http4s/http4s#7474] [General]
Continuing [a discussion that the bots closed](https://github.com/http4s/http4s/pull/7460#issuecomment-2174414534)

> This is going to be messy:
>
> - In Early Semver, we either have to drop Scala Native 0.4 in a patch release or make a breaking change that ripples the entire downstream ecosystem.
> - We can postpone that decision for a while by sticking to Cats 2.11 and similar, but that won't be tenable in the long term as we need new features from the base libraries.
>
> I think the least of all evils is just to proceed in a patch whenever it's green.

@satorg replied:

> But it is not just about Cats, isn't it? I mean, CE3 and FS2 are coming through similar tough choices. Moreover, I guess http4s cannot be released for Native 0.5 until the both upstream libraries go live with that – is that correct?
> 
> Besides, as far as I can see it, Cats is not going to receive updates for Native 0.4 beyond its v2.11. So sticking to Cats 2.11 means sticking to it forever.... On the second thought, however, there still could happen some kind of "emergency" patches to Cats for Native 0.4 that could be forked out of the v2.11 tag...
> 
> So if the goal is to let http4s 0.23 continue receiving patches against Native 0.4 while moving on with Native 0.5, I can think of two possible ways to accomplish that:
> 
> 1. Release a new 0.23 version against Native 0.5 with a tag that creates enough room for possible patches to the Native 0.4 line. For example, the new version for Native 0.23 could be 0.23.50 – therefore if there are any patches to the current line to occur, they can take numbers 0.23.28 through 0.23.49. Yeah, it seems quite artificial and clumsy, but might work to some extent.
> 
> 2. Just kick off a brand new 0.24 line based on Native 0.5, while keep the 0.23 one sticking to Native 0.4.

**Comment by rossabaker** [comment]
Any release of 0.24, even if binary compatible, will be flagged as incompatible and cause immense toil for http4s-integrated libraries and all the apps that use them.  I strongly believe the next breaking release needs to be 1.0, and come at high value.  There are but a handful of Scala Native 0.4 users, and I don't think ought to be a motivating reason.

I'm not opposed to @satorg's "leave some breathing room" proposal, or just saying Scala Native support is more experimental anyway and giving 0.4 a well-advertised final hurrah.

**Comment by danicheg** [comment]
Perhaps I'm overlooking something stupidly simple, but we already do updates of Scala 3 in patch releases (Scala 3 isn't forward-compatible across distinct minor series). Why do we need to worry so much about the Scala Native approach? Also, maybe I'm too biased towards auxiliary to JVM platforms, but complicating our tough lives should come with strong reasoning.

**Reply by danicheg** [reply]
Mehhh, to be precise, we _were_ updating Scala 3 in patch releases, but for now, we are sticking with Scala `3.3.x`. But this barely affects my sentiments on the matter 😄

**Reply by rossabaker** [reply]
I tried to stick to Scala 3.0.x, and 3.1.x, and so on, but there was no coherent community plan, and as soon as any dependency bumped, our hand was forced.  It took until 3.3.x before there was wide agreement on an LTS strategy.  Whenever a dependency moves past 3.3.x, we'll have a similar dilemma.

The main difference with Scala 3 is that it's supposed to be easy to upgrade the compiler in an app, and all libraries that already published for Scala 3 should remain usable.   A Scala Native bump requires lining up a new version stack, plus whatever breaking changes it made.

**Reply by danicheg** [reply]
There should probably be a collegial decision, similar to how we handle Scala 3 updates. Pinning updates of our upstream doesn't seem like a viable solution to me.
